### ISSUE TITLE
trial.py script fails on Windows with Python 2.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2489]
|<img alt="arkanes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| arkanes reported|
|-|-|
|Trac ID|trac#2489|
|Type|defect|
|Created|2007-02-26 19:25:10Z|

Python 2.5 made a change in the default items on sys.path - instead of the current working directory, it's the directory of the running script.

This means that the C:\Python25\Scripts\trial.py helper script doesn't work as expected unless your local directory is in PYTHONPATH for other reasons (like when running the Twisted test suite, for example).

The actual failure was very hard to diagnose, because it happens quite high up the callstack and results in an a cascade failure that doesn't manifest until much farther down, somewhere in t.p.modules. When t.trial.runner fails to find the module, it uses a fallback which ends up identifying the !__init!__.py file of the module, rather than the module itself. I've attached a copy of the actual traceback that happens with a stock trial.py, and another with an enhanced traceback from cgitb. If you look at the enhanced traceback, you'll see that the module is shown as ```<module '__init__' from 'testcopy\__init__.pyc'>```, rather than ```<module 'testcopy' from 'testcopy\__init__.pyc'>```. 

This is the cascading error from t.trial.runner.filenameToModule, which falls back on _importFromFile(), which imports the module but does so with the file name rather than the package name. 

There's an easy and dirty fix - modifying the trial.py runnner to add os.curdir to the path. Fixing filenameToModule to create the module correctly might also work doesn't seem to be enough - the correct module name is returned, but it's still not insterted into the import cache.

Attachments:

* [importerrors.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d89/d892229a4b224bd077941152ae8f73836f5066fc/bb0ace106f5724712f931db9c876d7c107c8df88.txt) (3855 bytes) - added by arkanes on 2007-02-26 19:25:40Z - Standard traceback
* [importerrors_cgitb.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d89/d892229a4b224bd077941152ae8f73836f5066fc/05885f343e4dd14fc80058b25f367a4bc37b22bc.txt) (12675 bytes) - added by arkanes on 2007-02-26 19:42:33Z - cgitb traceback, last time I promise

<details><summary>Searchable metadata</summary>

```
trac-id__2489 2489
type__defect defect
reporter__arkanes arkanes
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__trial trial
keywords__ 
time__1172517910000000 1172517910000000
changetime__1309878098000000 1309878098000000
version__None None
owner__jml jml
cc__teratorn cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

reformatting non-markup

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

importerrors_cgitb.txt looks pretty short.  I think you attached the wrong file?

**Comment by twisted-trac** [comment]
|<img alt="arkanes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|arkanes commented|
|-|-|

Replying to [exarkun](#note_2):
> importerrors_cgitb.txt looks pretty short.  I think you attached the wrong file?

Yes, sorry. And then I attached one run against the wrong version of runners.py. This one should be accurate.

In looking for a "clean" fix I'm not sure that the fallback code in runners.py ever worked. It creates a module by filename, but imp.load_source doesn't add the module to sys.modules or to sys.path_importer_cache, so any use of the fallback code (because of the Pyhon 2.5 change or not) will end up raising this same error.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_4"></a>|@zseil commented|
|-|-|

Note that the trial script already uses the "easy and dirty fix" on all platforms
that have os.getuid if it isn't ran by root.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

[#2005](https://github.com/twisted/twisted/issues/2005) and [#2507](https://github.com/twisted/twisted/issues/2507) are probably related to this.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank commented|
|-|-|

I'll just note that when a program is started like this in current directory `C:\temp`,
```
C:\Python27\python.exe C:\Python27\Scripts\somescript.py
```
, both `C:\temp` and `C:\Python27\Scripts` are in `sys.path`

When started like this:
```
C:\Python27\Scripts\somescript.py
```
, `C:\temp` is not in `sys.path` (but `C:\Python27\Scripts` is).

In both cases, `os.getcwd()` returns `C:\temp`.

If you run trial with the explicit `python.exe`, and remove the `getuid` security feature (?) from the preamble (always insert the current directory), trial seems to work pretty well on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_8"></a>|@zseil set status to `closed`|
|-|-|

This was fixed as part of [#3526](https://github.com/twisted/twisted/issues/3526). `cwd` is now prepended to `sys.path` in `trial` and `twistd` scripts, both when installed and when running from a checkout.
